src/Artifacts.jl

```diff
@@ -0,0 +1,594 @@
+returns the `Platform` object that this entry specifies.  Returns `nothing` on error.
+"""
+function unpack_platform(entry::Dict, name::String, artifacts_toml::String)
+    if !haskey(entry, "os")
+        @warn("Invalid Artifacts.toml at '$(artifacts_toml)': platform-specific artifact entry '$name' missing 'os' key")
+        return nothing
+    end
+
+    if !haskey(entry, "arch")
+        @warn("Invalid Artfiact.toml at '$(artifacts_toml)': platform-specific artifact entrty '$name' missing 'arch' key")
+        return nothing
+    end
+
+    # Helpers to pull out `Symbol`s and `VersionNumber`s while preserving `nothing`.
+    nosym(x::Nothing) = nothing
+    nosym(x) = Symbol(lowercase(x))
+    nover(x::Nothing) = nothing
+    nover(x) = VersionNumber(x)
+
+    # First, extract OS; we need to build a mapping here
+    os_map = Dict(
+        "windows" => Windows,
+        "macos" => MacOS,
+        "freebsd" => FreeBSD,
+        "linux" => Linux,
+    )
+    P = get(os_map, lowercase(entry["os"]), UnknownPlatform)
+
+    # Next, architecture, libc, libgfortran version and cxxabi (if given)
+    arch = nosym(get(entry, "arch", nothing))
+    libc = nosym(get(entry, "libc", nothing))
+    libgfortran_version = nover(get(entry, "libgfortran_version", nothing))
+    libstdcxx_version = nover(get(entry, "libstdcxx_version", nothing))
+    cxxstring_abi = nosym(get(entry, "cxxstring_abi", nothing))
+
+    # Construct the actual Platform object
+    return P(arch;
+        libc=libc,
+        compiler_abi=CompilerABI(
+            libgfortran_version=libgfortran_version,
+            libstdcxx_version=libstdcxx_version,
+            cxxstring_abi=cxxstring_abi
+        ),
+    )
+end
+
+function pack_platform!(meta::Dict, p::Platform)
+    os_map = Dict(
+        Windows => "windows",
+        MacOS => "macos",
+        FreeBSD => "freebsd",
+        Linux => "linux",
+    )
+    meta["os"] = os_map[typeof(p)]
+    meta["arch"] = string(arch(p))
+    if libc(p) != nothing
+        meta["libc"] = string(libc(p))
+    end
+    if libgfortran_version(p) != nothing
+        meta["libgfortran_version"] = string(libgfortran_version(p))
+    end
+    if libstdcxx_version(p) != nothing
+        meta["libstdcxx_version"] = string(libstdcxx_version(p))
+    end
+    if cxxstring_abi(p) != nothing
+        meta["cxxstring_abi"] = string(cxxstring_abi(p))
+    end
+end
+
+"""
+    artifact_meta(name::String, artifacts_toml::String;
+                  platform::Platform = platform_key_abi())
+
+Get metadata about a given artifact (identified by name) stored within the given
+`Artifacts.toml` file.  If the artifact is platform-specific, use `platform` to choose the
+most appropriate mapping.  If none is found, return `nothing`.
+"""
+function artifact_meta(name::String, artifacts_toml::String;
+                       platform::Platform = platform_key_abi())
+    if !isfile(artifacts_toml)
+        return nothing
+    end
```

Comment:
why not throw an error in these cases?

Reply:
Because it's completely normal to check for the existence of an artifact that may or may not exist; that shouldn't be an error, just a negative result.  An error should be reserved for some kind of exceptional circumstance, and I don't consider this an exceptional circumstance; there is no guarantee that a certain hash is bound within an `Artifacts.toml` file, because these files can be quite dynamic.